plugins/kv_oauth.ts

```diff
@@ -1,206 +1,107 @@
- * import { kvOAuthPlugin } from "https://deno.land/x/deno_kv_oauth@$VERSION/fresh.ts";
+ * import { createRoutes, createGitHubOAuthConfig } from "$fresh/plugins/kv_oauth.ts";
  * import manifest from "./fresh.gen.ts";
  *
  * await start(manifest, {
- *   plugins: [
- *      kvOAuthPlugin({
- *          github: createGitHubOAuth2Client(),
- *      })
- *   ]
+ *   plugins: [{
+ *     name: "kv-oauth",
+ *     routes: createRoutes(createGitHubOAuthConfig())
+ *   }]
  * });
  * ```
- */
-export function kvOAuthPlugin<
-  const TProviders extends Record<string, OAuth2Client>,
->(providers: TProviders): Plugin<TProviders>;
-
-/**
- * This creates handlers for the following routes:
- * - `GET /oauth/signin` for the sign-in page
- * - `GET /oauth/callback` for the callback page
- * - `GET /oauth/signout` for the sign-out page
  *
+ * @example Multiple providers
  * ```ts
  * // main.ts
  * import { start } from "$fresh/server.ts";
- * import kvOAuthPlugin from "$fresh/plugins/kv_oauth.ts";
- * import { createGitHubOAuth2Client } from "https://deno.land/x/deno_kv_oauth@$VERSION/mod.ts";
+ * import {
+ *   createRoutes,
+ *   createGitHubOAuthConfig,
+ *   createGoogleOAuthConfig,
+ * } from "$fresh/plugins/kv_oauth.ts";
  * import manifest from "./fresh.gen.ts";
  *
  * await start(manifest, {
- *   plugins: [
- *     kvOAuthPlugin(createGitHubOAuth2Client())
- *   ]
+ *   plugins: [{
+ *     name: "kv-oauth-github",
+ *     routes: createRoutes(createGitHubOAuthConfig(), "/oauth/github"),
+ *   }, {
+ *     name: "kv-oauth-google",
+ *     routes: createRoutes(createGoogleOAuthConfig(), "/oauth/google"),
+ *   }]
  * });
  * ```
  */
-export default function kvOAuthPlugin<
-  const TProviders extends Record<string, OAuth2Client>,
->(
-  ...args: [
-    oauth2Client: OAuth2Client,
-    options?: KvOAuthPluginOptions,
-  ] | [
-    providers: TProviders,
-  ]
-): Plugin<Record<keyof TProviders, unknown>> {
-  const routes: Plugin["routes"] = [];
-
-  if (args.length >= 3 || args.length <= 0) {
-    throw new Error(
-      `Unable to initialise kv-oauth plugin with. Expected 1-2 arguments, got ${args.length}.`,
-    );
-  }
-
-  const [providersOrOAuth2Client] = args;
-
-  if (providersOrOAuth2Client instanceof OAuth2Client) {
-    const [_, options] = args;
-    routes.push(
-      {
-        path: options?.signInPath ?? "/oauth/signin",
-        handler: async (req) => await signIn(req, providersOrOAuth2Client),
-      },
-      {
-        path: options?.callbackPath ?? "/oauth/callback",
-        handler: async (req) => {
-          // Return object also includes `accessToken` and `sessionId` properties.
-          const { response } = await handleCallback(
-            req,
-            providersOrOAuth2Client,
-          );
-          return response;
-        },
+export function createRoutes(
+  oauthConfig: OAuth2ClientConfig,
+  /** Parent path for sign-in, callback and sign-out pages. */
+  oauthPath = "/oauth",
+): PluginRoute[] {
```

Comment:
**suggestion:** would it make sense to convert this into a tuple rather than having an array?
this way, we could simply pass `oauthPath` as a generic to infer the paths, if needed:   

```tsx
export function createRoutes<T extends string = "/oauth">( // <-- `/oauth` would be needed so that we can infer the default
  oauthConfig: OAuth2ClientConfig,
  /** Parent path for sign-in, callback and sign-out pages. */
  oauthPath: T = "/oauth" as T,
): [
  { path: `${T}/signin`; handler: Handler }, // <-- each of these would still satisfy `PluginRoute`
  { path: `${T}/callback`; handler: Handler },
  { path: `${T}/signout`; handler: Handler },
] {
  return [
    {
      path: `${oauthPath}/signin`,
      handler: async (req) => await signIn(req, oauthConfig),
    },
    {
      path: `${oauthPath}/callback`,
      handler: async (req) => {
        // Return object also includes `accessToken` and `sessionId` properties.
        const { response } = await handleCallback(
          req,
          oauthConfig,
        );
        return response;
      },
    },
    {
      path: `${oauthPath}/signout`,
      handler: signOut,
    },
  ];
}

const routes = createRoutes(createGitHubOAuthConfig()); // <-- hovering over `routes` would already give us some nice intellisense

type Routes = typeof routes; // <-- this would let us assign it as a type
type Route = typeof routes[number]["path"]; // <-- this would give us a union type consisting of `"/oauth/signin" | "/oauth/callback" | "/oauth/signout"`
```

wdyt?

Reply:
I like this idea. I was originally going to try diving into the idea of making each array element a constant but I think using tuples is much smarter here.

Then if we want to update the `InferOAuthProviders` type we can update it to something like this to make it as easy as possible to get a typed route:
```ts
type InferOAuthProviders<T extends PluginRoute[]> = T[number]["path"];
```

Reply:
The one thing I will add:
> each of these would still satisfy `PluginRoute`

Do we want to rely on the explicit return for this type satisfaction or add a `satisfies PluginRoute[]` to the actual array being returned?
```ts
return [
	...
] satisfies PluginRoute[];
```

Reply:
Yeah, I was thinking that too.

Reply:
> Do we want to rely on the explicit return for this type satisfaction or add a satisfies PluginRoute[] to the actual array being returned?

thought about that as well, yet i wasn't able to appease ts.  
would absolutely prefer the usage of `satisfies PluginRoute[]`.

Reply:
I went with the approach that explicitly defines route paths. It provides greater detail.